Support LTO Builds

andyp1per opened the pull request:
Examples work, but not sure much else

andyp1per committed:
AP_HAL_ChibiOS: mark wrapped functions as used

andyp1per committed:
waf: add LTO flags

tpwrules commented:
I had investigated this in the past and done some work to fix related compiler errors but I wasn't sure if it was possible. In particular I'm not sure wrapped functions are safe to use, even if they are marked as used? IIUC there is a [GCC bug](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=88643) where it can resolve functions before they can get wrapped, although it looks like maybe there's a fix there?

I think we should kill the wrapped functions ultimately though.

There's also the problem that `-O2` boards now optimize too well and overflow the flash.

andyp1per commented:
@tpwrules great! Do you want to collaborate? The examples work for me, so clearly not all is broken - I am hoping its something simple. Apparently all the ChibiOS examples are built with LTO so there should be good underlying support.

fossuav-admin committed:
waf: use gcc-ar and disable builtins for wrapped libc under LTO

Plain ar omits LTO symbols from the archive index, so library members are
dropped at link (undefined references to AP_Gripper, AC_PosControl, etc).
Use the plugin-aware gcc-ar. Also hide the --wrap'd libc builtins from the
optimizer with -fno-builtin so LTO does not resolve them before the linker's
--wrap can redirect them, which otherwise pulls newlib back in and bypasses
the custom allocator.

fossuav-admin committed:
AP_InternalError: mark AP_stack_overflow used for LTO

Referenced only from ChibiOS C (stm32_util.c), so LTO drops it otherwise.

fossuav-admin committed:
AP_HAL_ChibiOS: mark dso_handle, scanf and malloc wraps used for LTO

These are referenced only from outside the LTO unit (the C++ runtime and the
linker --wrap), so LTO garbage-collects them without the used attribute.
